src/libexpr/eval.cc

```diff
@@ -599,21 +600,45 @@ void EvalState::allowAndSetStorePathString(const StorePath & storePath, Value &
     mkStorePathString(storePath, v);
 }
 
-void EvalState::checkURI(const std::string & uri)
+inline static bool isJustSchemePrefix(std::string_view prefix)
 {
-    if (!evalSettings.restrictEval) return;
+    return
+        !prefix.empty()
+        && prefix[prefix.size() - 1] == ':'
+        && isValidSchemeName(prefix.substr(0, prefix.size() - 1));
+}
 
+bool isAllowedURI(std::string_view uri, const Strings & allowedUris)
+{
     /* 'uri' should be equal to a prefix, or in a subdirectory of a
        prefix. Thus, the prefix https://github.co does not permit
-       access to https://github.com. Note: this allows 'http://' and
-       'https://' as prefixes for any http/https URI. */
-    for (auto & prefix : evalSettings.allowedUris.get())
-        if (uri == prefix ||
-            (uri.size() > prefix.size()
-            && prefix.size() > 0
-            && hasPrefix(uri, prefix)
-            && (prefix[prefix.size() - 1] == '/' || uri[prefix.size()] == '/')))
-            return;
+       access to https://github.com. */
+    for (auto & prefix : allowedUris) {
+        if (uri == prefix
```

Comment:
This seems like a special case of `hasPrefix(uri, prefix)`, and we could change the size check to `>=`. Feel free to ignore though, I don't think it matters much.

Reply:
It is not, because of the extra conditions that either
 - the prefix ends at a directory separate
 - or the prefix is a scheme prefix

The unit tests will catch this mistake, so I think it's ok not to leave a comment about something that isn't.